Progress Component

avelinet opened the pull request:
This PR adds the Progress component

Status: **Ready for Review**
Reviewers: @kpeatt @ry5n @jeffkamo @cole-sanderson @mlworks @nastiatikk 
JIRA: **[CSOPS-515](https://mobify.atlassian.net/browse/CSOPS-843)**

## Minor work todo
- [x] BLOCKER: Decide on less ambiguous name: `c-progress-steps`, `c-progress-stepped`, other
- [x] Consider renaming some sub-components

## Final review
- [x] +1 @ry5n
- [x] +1 @jeffkamo

nastiatikk commented:
@avelinet What do you think about tweaking a little background color for complete and current steps? They almost don't differ from each other and you can't see current step right away.

Thought it's just a performance note, but I believe that performance should show you very clear what is where.

We could tweak background like this:

$progress__current-bg-color: darken($progress__item-bg-color, 10%) !default;
$progress__completed-bg-color: darken($progress__item-bg-color, 20%) !default;

kpeatt commented on dist/components/progress/_progress.scss:
```diff
@@ -0,0 +1,119 @@
+$progress__divider-border-color: hsl(360, 100%, 100%) !default;
+
+
+.c-progress {
```
I feel like this should be a red flag when we build components.

jeffkamo replied:
I assume you mean when a class like this is empty

ry5n replied:
I don’t think an empty root is a red flag. In this case, it’s now a moot point.

kpeatt commented on tests/visual/components/progress/index.html:
```diff
@@ -0,0 +1,126 @@
+
+                        <div class="c-progress-step__item c--completed">
+                            <span class="c-progress-step__index">Step 1</span>
+                            <div class="c-progress-step__label">The Ingredients</div>
```
Why is index a `span` and label a `div`?

avelinet replied:
No real reason. They could both be either or depending on the design. I updated the dust file to be spans for consistency.

ry5n commented on dist/components/progress/_progress.scss:
```diff
@@ -0,0 +1,116 @@
+
+$progress__item-bg-color: hsl(0, 0%, 95%) !default;
+$progress__item-padding: if(variable-exists(v-space), $v-space, 10px) !default;
+$progress__stage-layer: if(variable-exists(stage-layer), $stage-layer, 10) !default;
```
Suggest changing this to `$progress__layer`.

ry5n commented on dist/components/progress/_progress.scss:
```diff
@@ -0,0 +1,116 @@
+$progress__divider-width: 20px !default; // 1
+$progress__divider-border-thickness: 2px !default;
+$progress__divider-border-color: hsl(360, 100%, 100%) !default;
+
```
These are great configurable vars: the CSS for building this is complex, so config vars make if very easy to customize.

ry5n commented on dist/components/progress/_progress.scss:
```diff
@@ -0,0 +1,116 @@
+// Progress: Steps
+// ---
+
+.c-progress__step-item {
```
Could just be `c-progress__step` or `c-progress-steps__item`.

jeffkamo replied:
For the record, the "root" component class name should not change for the sake of a one-off sub-component... That would make it non-self documenting.

ry5n commented on dist/components/progress/_progress.scss:
```diff
@@ -0,0 +1,116 @@
+
+.c-progress__inbetween,
+%c-progress__inbetween {
+}
```
What is this for?

ry5n commented on dist/components/progress/progress.js:
```diff
@@ -0,0 +1,38 @@
+        "maxStep": 0
+        "currentStep": 0,
+
+        "items": [
```
This is mock data, correct? It doesn’t match the schema though (here, "items", there, "steps").

ry5n commented on dist/templates/progress/_progress.dust:
```diff
@@ -0,0 +1,32 @@
+
+{!}
+
+<div role="progressbar" aria-valuetext="Step {currentStep} of {maxStep}" aria-valuenow="{currentStep}">
```
My intuition says the root node should have the root class. Looks like it was like that and got refactored out. Sounds like we need to discuss which oddness is worse: root class not on the root node, or root class with no styles. My feeling is root class should always be on the root node, and an empty class selector is fine here and there.

ry5n commented:
Solid work. Basically it’s some nitpicks about naming choices, whether or not we move the root class back to the root node, and otherwise +1.

avelinet commented:
@ry5n I've incorporated your feedback, thanks!

jeffkamo commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,108 @@
+// Progress: Index
+// ---
+
+.c-progress-steps__index {
```
Is there a need for these? I suppose they're here just to demonstrate that they are available for customization?

Perhaps some comments to document that would be in order.

avelinet replied:
Yes this allows completed steps to be a different color

jeffkamo commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,108 @@
+// Progress: Label
+// ---
+
+.c-progress-steps__label {
```
Same for this.

avelinet replied:
This is for styling of Step index and Step labels. Perhaps they don't need to be included in the component css.

jeffkamo commented:
Wait, why are there files in `dist/templates/`? Are those outdated files? Pretty sure there isn't a template directory in dist, or am I mistaken?

jeffkamo commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,108 @@
+//
+// Helps show users where they exist within a multiple-page workflow, such as a
+// checkout flow.
+
```
A little nitpick: there should be two spaces above headings, not one.

jeffkamo commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,108 @@
+$progress-steps__completed-bg-color: darken($progress-steps__item-bg-color, 20%) !default;
+
+// Arrow Divider Variables
+// 1. These two variables together determine the angle of the arrow dividers
```
There should be an additional space above this ordered list to prevent markdown from interpreting these two lines as a single paragraph.

jeffkamo replied:
Hmm, I wonder if the line notes should go with the section heading, instead of this sub-sub heading?

Thoughts @ry5n?

jeffkamo commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,108 @@
+// ---
+
+.c-progress-steps__item {
+    position: relative; // For the absolutely positioned arrows
```
Why isn't this comment a line comment?

jeffkamo commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,108 @@
+
+    background-color: $progress-steps__item-bg-color;
+
+    // Adjust padding to account for the arrows
```
same here, this could be a line comment

jeffkamo commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,108 @@
+
+    // background color spacing
+    &::before {
+        margin-left: $progress-steps__divider-border-thickness; // offset to create the appearance of a border
```
same again, line comment

avelinet commented:
@jeffkamo The template files in dist are out of date. After all this thing dates back to December! I've removed them.

ry5n commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,111 @@
+// ---
+//
+// Notes:
+// [1] For the absolutely positioned arrows
```
Comment formatting should follow the mobify code style section on comments, so:

```scss
Notes:
1. Hide this for some reason

.c-code {
   display: none; // 1
}
```

ry5n commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,111 @@
+// [3] Coloured triangle
+// [4] Background color spacing
+// [5] Offset to create the appearance of a border
+
```
Comment formatting should follow the mobify code style section on comments, so:

```scss
Notes:
1. Hide this for some reason

.c-code {
   display: none; // 1
}
```

jeffkamo commented:
@avelinet I think you removed the wrong template files. The root `/template` directory is fine as is, it's the `/dist/templates` files that I think need to be removed

ry5n commented on dist/components/progress-steps/_progress-steps.dust:
```diff
@@ -0,0 +1,19 @@
+
+{!}
+
+{#progressStep}
```
I don’t think we want this wrapping key: it should be the responsibility of the developer to make sure the correct keys are available in the template without setting a context here, I think.

ry5n commented:
Not sure about what’s happening with the templates directory, but once that’s resolved we can merge. I still think the top-level node should have the root class, and the dust template issue I mentioned above, but we could make those new issues for later. +1 on merge.

nastiatikk commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    // 4
+    &::before {
```
I find section comments like `// 3` and `// 4` to be confusing. 

Comments are used to give a title or to describe something unclear. And even though I know what `border-left` is doing in this case I need to scroll up and to check what is `// 3` cause comment also means warning.

Plus it's not easy to read. I'd prefer to have `Coloured triangle` as a title of `&::after`. Or not having it at all.

nastiatikk commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,111 @@
+    // 3
+    &::after {
+        border-left: $progress-steps__divider-width solid $progress-steps__item-bg-color;
+    }
```
I find section comments like `// 3` and `// 4` to be confusing. 

Comments are used to give a title or to describe something unclear. And even though I know what `border-left` is doing in this case I need to scroll up and to check what is `// 3` cause comment also means warning.

Plus it's not easy to read. I'd prefer to have `Coloured triangle` as a title of `&::after`. Or not having it at all.

ry5n replied:
Agree. Always prefer comments placed as close as possible to the source code. Footnote-style comments should only be used when adding lots of comments to a style block that would clutter the code. The referenced comment should be placed here instead of the 3.

avelinet replied:
Ok well now I'm just confused about how we want to do comments https://github.com/mobify/stencil/pull/108#discussion_r26686833

ry5n replied:
I guess @jeffkamo and I disagree…

nastiatikk replied:
I see it clear
Have `Coloured triangle` instead of `// 3`
Have `Triangle border` instead of `// 4`

ry5n replied:
Update: I should clarify that our official code style says never to use line comments; I guess I hadn’t done a close enough reading. So the current state is correct. @nastiatikk we should open an issue on mobify-code-style if we want to allow line comments sometimes.

jeffkamo replied:
Hmm, I don't think our comment documentation addresses the difference between section (sub)headings and comments. That's a something that's probably worth mentioning...

I'm okay-ish with adding comments that are effectively subheadings. It's the comments that are that are used to describe one or more specific properties that need to be formatted as described in our documentation.

jeffkamo replied:
If it's a matter of keeping documentation comments close to the code it's describing, then maybe that's a sign that a block is too massive and should be broken out into its own section.

Much like we sometimes do with theme styles: instead of including a ton of `&.c--modifier` classes and blocks in a single block, it's broken out under a "Themes" section.

Could maybe do something similar with these arrows.

nastiatikk commented on dist/components/progress-steps/_progress-steps.scss:
```diff
@@ -0,0 +1,111 @@
+// 1. For the absolutely positioned arrows
+// 2. Adjust padding to account for the arrows
+// 3. Coloured triangle
+// 4. Background color spacing
```
Probably rename `Background color spacing` to `Triangle border`?

avelinet commented:
@jeffkamo @nastiatikk @ry5n, appreciate all the feedback on this. I reverted the removal of the wrong template files and removed the correct ones. Asides from issues with comment placement and semantics, discussions on root classes and dust templating wrapping keys, all of which require further discussion until the 🐄s come home... Are there any blockers remaining or can this be merged in?

ry5n commented:
Nope this is good to merge, the rest are bikesheds!

jeffkamo commented:
Aah, not quite. Aveline and I realized that the comments are actually rather confusing. We're going to clarify those after lunch, and at that point we'll most likely be good to merge!

jeffkamo commented:
+1'ed!
